src/main/kotlin/com/terraformation/backend/nursery/db/BatchStore.kt

```diff
@@ -264,6 +285,142 @@ class BatchStore(
+                    // available quantities. However, if the user is entering data about an older
+                    // withdrawal, and they have explicitly edited the available quantities more
+                    // recently than the withdrawal, just record the withdrawal without subtracting
+                    // it from inventory.
+                    //
+                    // We figure this out inside the retry loop because the latest observed time
+                    // might have been updated by whatever operation caused our version number to
+                    // be out of date.
+                    val latestObservedDate =
+                        LocalDate.ofInstant(batch.latestObservedTime!!, ZoneOffset.UTC)
+                    val withdrawalIsNewerThanObservation =
+                        withdrawal.withdrawnDate >= latestObservedDate
+
+                    if (withdrawalIsNewerThanObservation) {
+                      val newGerminatingQuantity =
+                          batch.germinatingQuantity!! - batchWithdrawal.germinatingQuantityWithdrawn
+                      val newNotReadyQuantity =
+                          batch.notReadyQuantity!! - batchWithdrawal.notReadyQuantityWithdrawn
+                      val newReadyQuantity =
+                          batch.readyQuantity!! - batchWithdrawal.readyQuantityWithdrawn
+
+                      if (newGerminatingQuantity < 0 ||
+                          newNotReadyQuantity < 0 ||
+                          newReadyQuantity < 0) {
+                        throw BatchInventoryInsufficientException(batchId)
+                      }
+
+                      // Update the batch's quantities to reflect the withdrawal. If another thread
+                      // has updated the batch since we fetched it above, the version number won't
+                      // match and BatchStaleException will be thrown, which will cause us to try
+                      // again.
+                      updateQuantities(
+                          batchId,
+                          batch.version!!,
+                          newGerminatingQuantity,
+                          newNotReadyQuantity,
+                          newReadyQuantity,
+                          BatchQuantityHistoryType.Computed,
+                          withdrawalId)
+                    } else {
+                      updateVersionedBatch(batchId, batch.version!!) {
+                        // Just update the version and modified user/timestamp (need a set() call
+                        // here so the query object is of the correct jOOQ type)
+                        it.set(emptyMap<Any, Any>())
+                      }
+                    }
+
+                    batchWithdrawalsDao.insert(batchWithdrawalsRow)
+
+                    succeeded = true
+                  } catch (e: BatchStaleException) {
+                    if (retriesRemaining > 0) {
```

Comment:
You can think of this as the optimistic-locking equivalent of doing a `SELECT FOR UPDATE` and blocking until the competing transaction is finished. If two withdrawals are happening at the same time (and there's enough inventory for both), they should both succeed without the client needing to know there was a collision.

Oh, I guess maybe the thing that makes this unclear is that in this case, unlike the case of editing a batch's quantities, the client is not passing in any version numbers. Semantically, "change the remaining quantity to 16" and "change the remaining quantity to 18" are conflicting operations, but "withdraw 3 seedlings" and "withdraw 4 seedlings" don't conflict, so there's no need for client-side conflict resolution in the latter case.

Reply:
Thanks for the explanation - this makes total sense.

I will add a note in the FE task to refresh the view prior to showing errors - in case the withdrawal failed because another withdrawal happened same time, and current request led to insufficient inventory.